src/core/CompUnit/Repository/Installation.pm

```diff
@@ -372,83 +372,68 @@ sub MAIN(:$name is copy, :$auth, :$ver, *@, *%) {
     }
 
     method script($file, :$name!, :$auth, :$ver) {
-        my $prefix = self.prefix;
-        my $lookup = $prefix.add('short').add(nqp::sha1($file));
-        return unless $lookup.e;
-
-        # Scripts using this interface could only have been installed long after the introduction of
-        # repo version 1, so we don't have to care about very old repos in this method.
-        my @dists = $lookup.dir.map({
-                my ($ver, $auth, $api, $resource-id) = $_.slurp.split("\n");
-                $resource-id ||= self!read-dist($_.basename)<files>{$file};
-                (id => $_.basename, ver => Version.new( $ver || 0 ), :$auth, :$api, :$resource-id).hash
-            }).grep({
-                $_.<auth> ~~ $auth
-                and $_.<ver> ~~ $ver
-            });
-        for @dists.sort(*.<ver>).reverse {
-            return self!resources-dir.add($_<resource-id>);
-        }
+        my $matching-scripts := self.files($file, :$name, :$auth, :$ver);
+        my $scripts = $matching-scripts.map: { .<files>{$file} }
+        return $scripts.head;
     }
 
     method files($file, :$name!, :$auth, :$ver) {
-        my @candi;
-        my $prefix = self.prefix;
-        my $lookup = $prefix.add('short').add(nqp::sha1($name));
-        if $lookup.e {
-            my $repo-version = self!repository-version;
-            my @dists = $repo-version < 1
-                ?? $lookup.lines.unique.map({
-                        self!read-dist($_)
-                    })
-                !! $lookup.dir.map({
-                        my ($ver, $auth, $api) = $_.slurp.split("\n");
-                        (id => $_.basename, ver => Version.new( $ver || 0 ), auth => $auth, api => $api).hash
-                    });
-            for @dists.grep({$_<auth> ~~ $auth and $_<ver> ~~ $ver}) -> $dist is copy {
-                $dist = self!read-dist($dist<id>) if $repo-version >= 1;
-                with $dist<files>{$file} {
-                    my $candi = %$dist;
-                    $candi<files>{$file} = self!resources-dir.add($candi<files>{$file});
-                    @candi.push: $candi;
-                }
+        my $spec = CompUnit::DependencySpecification.new(
+            :short-name($name),
+            :auth-matcher($auth || True),
+            :version-matcher($ver // '*'),
+        );
+
+        with self!matching-dist($spec) {
+            my $dist-ids       := $_.map: { .hash.keys[0] }
+            my $matches-spec   := $dist-ids.map: { self!read-dist($_) }
+            my $matches-file   := $matches-spec.grep: { .<files>{$file} }
+
+            $matches-file.map: -> $candi is copy {
```

Comment:
I think an explicit return would not hurt here as we're within a block and the missing semicolon after the closure doesn't have any effect.